.pfnci/windows/test.ps1

```diff
@@ -0,0 +1,110 @@
+Param(
+    [String]$cuda,
+    [String]$python,
+    [String]$test
+)
+
+$ErrorActionPreference = "Stop"
+. "$PSScriptRoot\_error_handler.ps1"
+
+. "$PSScriptRoot\_flexci.ps1"
+
+$cupy_kernel_cache_file = "cupy_kernel_cache_windows.zip"
+
+
+function DownloadCache {
+    pushd $Env:USERPROFILE
+    echo "Downloading kernel cache..."
+    gsutil -m -q cp gs://tmp-asia-pfn-public-ci/cupy-ci/$cupy_kernel_cache_file .
+    if (-not $?) {
+        echo "*** Kernel cache unavailable"
+    } else {
+        echo "Extracting kernel cache..."
+        RunOrDie 7z x $cupy_kernel_cache_file
+        rm $cupy_kernel_cache_file
+    }
+    popd
+}
+
+function UploadCache {
+    # Expires in 3 days, maximum 1 GB
+    echo "Trimming kernel cache..."
+    RunOrDie python .pfnci\trim_cupy_kernel_cache.py --expiry 259200 --max-size 1000000000 --rm
+
+    pushd $Env:USERPROFILE
+    # -mx=0 ... no compression
+    # -mtc=on ... preserve timestamp
+    echo "Compressing kernel cache..."
+    RunOrDie 7z a -tzip -mx=0 -mtc=on $cupy_kernel_cache_file .cupy
+    echo "Uploading kernel cache..."
+    RunOrDie gsutil -m -q cp $cupy_kernel_cache_file gs://tmp-asia-pfn-public-ci/cupy-ci/
+    popd
+}
+
+function Main {
+    # Setup environment
+    echo "Using CUDA $cuda and Python $python"
+    ActivateCUDA $cuda
+    ActivatePython $python
+
+    # Setup build environment variables
+    $Env:CUPY_NUM_BUILD_JOBS = "16"
+    $Env:CUPY_NVCC_GENERATE_CODE = "current"
+    echo "Environment:"
+    RunOrDie cmd.exe /C set
+
+    # Build
+    echo "Setting up test environment"
+    RunOrDie python -V
+    RunOrDie python -m pip install Cython scipy optuna
+    RunOrDie python -m pip freeze
+
+    echo "Building..."
+    RunOrDie python -m pip install -e ".[all,jenkins]" -vvv > cupy_build_log.txt
+
+    # Import test
+    echo "CuPy Configuration:"
+    RunOrDie python -c "import cupy; cupy.show_config()"
+
+    # Unit test
+    $pytest_opts = ""
+    if ($test -eq "build") {
+        return
+    } elseif ($test -eq "test") {
+        $pytest_opts = "$pytest_opts -m ""not slow"""
+    } elseif ($test -eq "slow") {
+        $pytest_opts = "$pytest_opts -m ""slow"""
+    } else {
+        throw "Unsupported test target: $target"
+    }
+
+    $use_cache = ($Env:CUPY_CI_CACHE_KERNEL -eq "1")
+
+    if ($use_cache) {
+        DownloadCache
+    }
+    echo "Running test..."
+    python -m pytest -rfEX $Env:PYTEST_OPTS tests > cupy_test_log.txt
+    if (-not $?) {
+        $test_retval = $LastExitCode
+    }
+    if ($use_cache) {
+        UploadCache
```

Comment:
It looks like a PR with errors will upload their invalid kernels here, forcing everyone to recompile?
I think we need to check for `test_retval` before uploading the cache.

I wonder if it's better to run the tests only once a day and then upload the cache from the master branch, rather than having all the tests uploading their cache.

Reply:
We can just address this in another PR, lets leave it as TODO :)